routes/User.js

```diff
@@ -0,0 +1,141 @@
+const express = require("express");
+const router = express.Router();
+const magic = require("../magic");
+const { User, Chapter, Module } = require("../models");
+const { Op } = require("sequelize");
+
+router.get("/", async (req, res) => {
+	const { uuid, xtz } = req.query;
+	console.log(uuid, xtz);
+
+	try {
+		const param = uuid ? { uuid } : { xtzAddress: xtz };
+
+		const user = await User.findOne({ where: param });
+
+		if (!user) {
+			throw new Error();
+		}
+
+		res.json(user);
+	} catch {
+		res.status(404).json({ error: "USER_NOT_FOUND" });
+	}
+});
+
+router.post("/", async (req, res) => {
+	const { xtzAddress } = req.body;
+
+	try {
+		let user = await User.findOne({ where: { xtzAddress } });
+
+		if (!user) user = await User.create({ xtzAddress });
+
+		return res.json(user);
+	} catch (err) {
+		return res.status(500).json(err.toString());
+	}
+});
+
+router.patch("/", async (req, res) => {
+	const { user: user_in_req, email, name } = req.body;
+
+	try {
+		let u = await User.findOne({
+			where: { [Op.or]: [{ email }, { name }] },
+		});
+		if (u) {
+			const err = u.email == email ? "EMAIL" : "NAME";
+
+			throw new Error(`${err}_ALREADY_USED`);
+		}
+		try {
+			console.log("user_in_req", user_in_req);
+			const user = await User.findOne({
+				where: {
+					uuid: user_in_req.uuid,
+					xtzAddress: user_in_req.xtzAddress,
+				},
+			});
+			console.log("user", user);
+			if (!user) {
+				throw new Error("USER_NOT_FOUND");
+			}
+
+			user.email = email;
+			user.name = name;
+
+			await user.save();
+
+			console.log("user saved.");
+
+			res.json(user);
+		} catch (err) {
+			res.status(404).json({ error: err.message });
+		}
+	} catch (err) {
+		res.status(400).json({ error: err.message });
+	}
+});
+
+router.post("/verify", async (req, res) => {
+	const { token } = req.body;
+	try {
+		const { email } = await magic.users.getMetadataByToken(token);
+		try {
+			const user = await User.findOne({ where: { email } });
+			if (!user) {
+				throw new Error("USER_NOT_FOUND");
+			}
+			user.verified = true;
+			await user.save();
+			return res.json(user);
+		} catch (err) {
+			res.status(404).json({ error: err.message });
+			return;
+		}
+	} catch (err) {
+		res.status(500).json({ error: err.message });
+	}
+});
+
+router.post("/progress", async (req, res) => {
```

Comment:
Please explain in depth what the progress endpoint is doing. Didn't understood it.

Reply:
Progress endpoint takes a user, module_num, and chapter_num.
It creates a relationship between the chapter with chapter_num n, and module_num m with the user - essentially marks that chapter as done for the user.